### DISCUSSION TITLE
Type fields

### DISCUSSION THREAD

**Discussion by wikoci** [payloadcms/payload#684] [Feature Requests & Ideas]
Feature : any Json type field ?

**Comment by jmikrut** [comment]
Hey @wikoci — technically, the `richText` field type stores its content as JSON in the database, but we don't have an exact `json` field as of now. But, this is a great idea to add. We would accept a PR for this, but otherwise I will convert this into a Feature Request discussion and we will report back to it ASAP!

**Comment by wikoci** [comment]
Thanks . waiting for this one to fully exploit the collections.

**Reply by balmacefa** [reply]
@jmikrut can you point/guide us how to add custom types? I can create a PR for the Json field..
Not sure if the design is build in a way that we can add plugins to create custom types too.

**Reply by jmikrut** [reply]
Hey @balmacefa — so, custom fields are done via overriding a _**built-in**_ field type. Payload field types, at their core, specify _**how data is saved in the database**_ - i.e. if a field has subfields, if it is an array-based field type, if it's text-based, number-based, etc.

So, when you define a custom field, you basically just choose an existing field type that corresponds to the type of data that your custom field needs to store, and then override whatever you need to override to customize it (validation, field React component, etc).

In this case, a JSON field would need to store data as a "Mixed" Mongoose schema type. Our existing rich text field _**does**_ use the "Mixed" schema type, but it would be fairly weird to specify a field as `richText` and then have it save JSON. So, for that reason, I think extending Payload's built-in field types with a new field type called `json` is the best bet here.

This needs to be done in quite a few places throughout the codebase, but it should be able to be done fairly easily.

What I'd do is search in the `/src` folder for some other built-in field type, like `email` for example, and then look at all of the places that the `email` field type functionality is located. Then in each of those places, build in the corresponding `json` field type functionality. 

You'll see that there are React components in the `/src/admin` folder that correspond to each of our built-in field types, and you'd need to create a JSON editor field type there as well (using some off-the-shelf JSON editor React component probably).

You should give this a shot! That would be great! PS - I'm in Discord actively and would be happy to pair with you to get you going. this would be a super great feature to add and I'd prioritize helping here accordingly.

**Reply by balmacefa** [reply]
I'll look into it. Yes is weird, there should be primitive types, so future changes won't impact implementations... `richText` type can change in the future, specially if other Database are planning to be supported.
I think this field/type definition can be expose on the payload config file, similar as collection.

There are npm packages to abstract db operation like typeorm, maintain simplicity as payload CMS does but support easily other/complex DB scenarios.

**Reply by jmikrut** [reply]
@balmacefa totally. We do plan to maintain an abstraction over _**any**_ lower-level ORM we end up using in the future for other databases, but I hear your point!

**Comment by DanRibbens** [comment]
Just an update on an old issue. We have a new discussion about this topic and are planning to implement it. https://github.com/payloadcms/payload/discussions/1292
